### PR TITLE
Optimise Aliases.get_canonical_element_exn

### PR DESCRIPTION
**Pull request by mshinwell**
This should be a slight improvement in efficiency (also it makes the code easier to read).

### CODE REVIEW

**Comment by mshinwell** [comment]
@xclerc do you think you could review this one please?  It is just a mechanical refactoring iirc.

File: middle_end/flambda2/types/env/aliases.ml [hunk 1]
```diff
@@ -982,89 +982,96 @@ let add ~binding_time_resolver ~binding_times_and_modes t
...
+   these situations to arise. *)
+
+let find_earliest_alias t ~canonical_element ~binding_times_and_modes
+    ~min_binding_time ~min_name_mode ~binding_time_resolver
+    ~coercion_from_canonical_to_element =
+  (* There used to be a shortcut that avoided consulting the aliases in the
+     common case that [element] is itself canonical and has no aliases, since
+     then it does not appear in [canonical_elements]. However, this shortcut was
+     broken: a canonical element *with* known aliases may still not appear in
+     [canonical_elements]. See tests/flambda2-aliases for a test that gave
```
Comment (xclerc) [inline middle_end/flambda2/types/env/aliases.ml:997]: (This is not related to this pull request, but I cannot find `tests/flambda2-aliases`.)

File: middle_end/flambda2/types/env/aliases.ml [hunk 2]
```diff
@@ -982,89 +982,96 @@ let add ~binding_time_resolver ~binding_times_and_modes t
...
-          names
+  match canonical t element with
+  | Is_canonical
+    when match Name_mode.compare_partial_order elt_name_mode min_name_mode with
+         | None -> false
+         | Some c -> c >= 0 ->
+    element
+  | Is_canonical | Alias_of_canonical _ -> (
+    let canonical_element, name_mode, coercion_from_canonical_to_element =
+      match canonical t element with
```
Comment (xclerc) [inline middle_end/flambda2/types/env/aliases.ml:1050]: (It is probably not necessary to evaluate `canonical t element` again.)

File: middle_end/flambda2/types/env/aliases.ml [hunk 3]
```diff
@@ -982,89 +982,96 @@ let add ~binding_time_resolver ~binding_times_and_modes t
...
-    then
-      Simple.with_coercion canonical_element coercion_from_canonical_to_element
-    else find_earliest ()
+    assert (not (Simple.has_coercion canonical_element));
+    match Name_mode.compare_partial_order name_mode min_name_mode with
+    | None ->
+      find_earliest_alias t ~canonical_element ~binding_times_and_modes
+        ~min_binding_time ~min_name_mode ~binding_time_resolver
+        ~coercion_from_canonical_to_element
+    | Some c ->
```
Comment (xclerc) [inline middle_end/flambda2/types/env/aliases.ml:1066]: (This `match` could use the same scheme as on line 1044,
perhaps even factoring it out to a utility function.)

**Review by xclerc — APPROVED** [review]

File: middle_end/flambda2/types/env/aliases.ml (hunk 1 above)
Comment (xclerc) [reply middle_end/flambda2/types/env/aliases.ml:997]: (Maybe it is https://github.com/ocaml-flambda/ocaml/blob/flambda2.0-stable/testsuite/tests/flambda2-aliases/test.ml.)

File: middle_end/flambda2/types/env/aliases.ml (hunk 3 above)
Comment (mshinwell) [reply middle_end/flambda2/types/env/aliases.ml:1066]: I think I'm going to leave this as-is for the moment.

File: middle_end/flambda2/types/env/aliases.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/flambda2/types/env/aliases.ml:997]: Thanks, I've added a CR for this test to be imported later.

**Comment by mshinwell** [comment]
@xclerc 's review will suffice for this.

**Pull request merged**